gnowledge/gstudio
# Curation Work Flow implemented

**Features of curation workflow**:
- Now moderation group hierarchy can be created.
- Hierarchy gets created with decided altnames.
- Only superuser will have access to create group.
- New simple template added to create any type of group.
- In models, status type choices will now have: `MODERATION` as one of the possible object-status. 
- Added moderation view, url and template to facilitate following moderation functionality:
	- Extended data-review for moderation
		- Currently, only `Approve` or `Publish` to next group functionality implemented.
		- Provision for *editorial-discussion* and *reject resource* made(under progress).
		- Showing total count of resources along with alternate moderator group name.
	- As resource gets into moderation, task gets created and status of resource changed to `MODERATION`.
	- When resource traverse through curation from one group to next mod group, task gets created.
	- At the end when resource gets final approved, it gets published to parent group.
		- Task intimation sent to both user and admin of group.
		- Status of resource updates to `PUBLISHED`.
- While resource is under moderation:
	- It's current status tracking functionality implemented.
		- Once redirected to this template, user will have detailed explanation of current status.
		- Message indicating confirmation to contributor that resource is uploaded successfully.
		- This is the template/functionality used to redirect user after upload into moderated group.
		- Shows the card of resource redirecting to user's group.
		- Also shows in which group resource currently lies. Along with remaining level of mederation to be traversed for the resource to get fully moderated.
		- On detail view of `under moderation` resource, `check current status of moderation` link is provided.
	- For looking/tracking no of resources are under moderation of parent group:
		- Button is provided `Moderation Status` to check status of all the resources at any levels of moderation.
		- This functionality is only available to super-user and admin and creator of top group.
		- In this view, all the resources will be visible as a cards. And on clicking of card gets you to above (`check status of resource`) view.
- Hiding group level appbar for moderation sub-groups.
- Action buttons on detail view modified for Moderated groups.
- Added `scss`/`css` to show MODERATION cards as `gray` colored.
- Disabled `group name` edit functionality

--

**Technicalities in implementation:**
- For the first time implemented views with latest approach of Class Based Views(CBV).
	- For creation of any type of group classes are implemented.
	- Implemented OOPS concepts like Inheritance, encapsulation.
	- Reduced re-writing redundant code with CBV.
	- Improved code readability and easy understanding.
	- Just import one class and all set of functionality will get shipped together.
<pre>
	class CreateGroup()
           |
	class CreateSubGroup(CreateGroup)
           |
        class ModeratedGroup(CreateSubGroup)
</pre>
- Wherever possible replaced group `_id` getting redundant code with one line function.
- `sync_existing_documents` modified to set `moderation_level` of groups other than `EDITABLE_MODERATED` to `-1`
- Using `file_list_tab` for listing of under moderation resources.
- Commented template level caching only for `header.html` for temporary.
- Added new item in `Moderation` in `GSTUDIO_TASK_TYPES` of `settings.py`.

--

**Not Implemented/Pending:**
- `Reject`/`Editorial Thread`.
- `Cross publish to another group`.
- Redirecting group edit to same (create) template.
- accessing all the sub-moderated group from parent group.
- adding/removing moderation levels (back-end view is ready).
- Currently, only `Files` are tested for the flow. Gradually all the GSystem's will fall under the flow.
- ...
- ...
- Possibly some other things can get into list. If so let me know.

--

**What to test**
- Create new moderated group with:
	- `edit_policy`: `EDITABLE_MODERATED`
	- `moderation_level`: `2`
- Try to upload resource into the group.
- Now moderate/approve resource from each of the group.
	- Currently, access it via group listing or via url
	- e.g: `<top_group_name>_mod/moderation` and `<top_group_name>_mod_mod/moderation`.
- In between check `Moderation status` functionality.
- Try to access resource from user's group (check for the warning at top).
- Check difference between under_moderation resources and other resources from list/card view.
